src/mtcp/mtcp_restart.c

```diff
@@ -850,7 +850,10 @@ unmap_memory_areas_and_restore_vdso(RestoreInfo *rinfo)
       mtcp_abort();
     }
     MTCP_ASSERT(vvar == vvarStart);
-    mtcp_memcpy(vvarStart, rinfo->vvarStart, vvarEnd - vvarStart);
+    // On i386, only the first 4096 bytes are readable. Reading beyond that
+    // results in a bus error.
+    // TODO: Find a way to automatically detect the readable bytes.
+    mtcp_memcpy(vvarStart, rinfo->vvarStart, 4096);//vvarEnd - vvarStart);
```

Comment:
@gc00, @karya0: I tried the `mprotect()` approach. It still fails on the memcpy. For reference here's what I tried:
```C
    MTCP_PRINTF("Before mprotect");
    mtcp_sys_mprotect(vvarStart, vvarEnd - vvarStart, PROT_READ);
    MTCP_PRINTF("After mprotect");
    mtcp_memcpy(vvarStart, rinfo->vvarStart, vvarEnd - vvarStart);
    //mtcp_memcpy(vvarStart, rinfo->vvarStart, 4096);//vvarEnd - vvarStart);
```